storm-client/src/jvm/org/apache/storm/topology/SimpleWindowPartitionCache.java

```diff
@@ -0,0 +1,191 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.storm.topology;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.locks.ReentrantLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A simple implementation that evicts the largest un-pinned entry from the cache. This works well
+ * for caching window partitions since the access pattern is mostly sequential scans.
+ */
+public class SimpleWindowPartitionCache<K, V> implements WindowPartitionCache<K, V> {
+    private static final Logger LOG = LoggerFactory.getLogger(SimpleWindowPartitionCache.class);
+
+    private final ConcurrentSkipListMap<K, V> map = new ConcurrentSkipListMap<>();
+    private final Map<K, Long> pinned = new HashMap<>();
+    private final long maximumSize;
+    private final RemovalListener<K, V> removalListener;
+    private final CacheLoader<K, V> cacheLoader;
+    private final ReentrantLock lock = new ReentrantLock(true);
+    private int size;
```

Comment:
`ConcurrentSkipList#size` is not a constant time operation and it may need to traverse all of them and its value also may change when size() is invoked.